yarnitor/background/worker.py

```diff
@@ -34,260 +38,390 @@
         self.name = name
         self.completed = completed
         self.failed = failed
         self.running = running
         self.total = total
 
+    def to_dict(self):
+        return self.__dict__
 
-class YarnApi(object):
-    """Collection of yarn endpoints we care about.
-    """
 
+class YARNHandler(object):
+    """Manages HTTP communication with a YARN ResourceManager (RM) to fetch
+    information about applications.
+
+    Parameters
+    ----------
+    host: str
+        Protocol, hostname, and optional port of the YARN RM
+    version: str, optional
+        Version of the YARN RM API
+
+    Attributes
+    ----------
+    base_url: dict
+        'host' and 'version' of the YARN RM
+    """
     def __init__(self, host, version="v1"):
-        self.host = host
-        self.version = version
+        self.base_url = {
+            'host': host,
+            'version': version
+        }
+
+    def get_url(self, path, **params):
+        """Issues an HTTP GET to the given path with the given parameters
+        and treats the response as JSON.
 
-    def get_url(self, url, **params):
-        final_url = "{host}/ws/{version}".format(**self.__dict__) + url
-        resp = requests.get(final_url, params)
+        Parameters
+        ----------
+        path: str
+            Path to append to the root YARN RM path
+        **params, dict
+            Query parameters to append to the YARN RM URL
+
+        Returns
+        -------
+        dict
+            JSON decoded response
+        """
+        # Handle HA redirects in a loop so that we can control how many
+        # we're willing to follow and don't blow out the stack with recursion.
+        for i in range(MAX_HA_REDIRECTS):
+            final_url = "{host}/ws/{version}/{path}".format(path=path, **self.base_url)
+            resp = requests.get(final_url, params)
+            resp.raise_for_status()
+            # YARN uses the Refresh header to indicate a change in the primary RM
+            ha_redirect = resp.headers.get('Refresh')
+            if ha_redirect is None:
+                break
+            else:
+                # Store the new host in the instance for the next request
+                # It comes in the form "3; url=http://newhost:port/same/path/as/request?args"
+                _, key = ha_redirect.split(';')
+                _, new_url = key.strip().split('url=')
+                parsed = urlparse(new_url.strip())
+                new_host = '{parsed.scheme}://{parsed.netloc}'.format(parsed=parsed)
+                self.base_url['host'] = new_host
+                logger.warn('YARN HA redirect, switching to URL: %s', new_host)
         return resp.json()
 
-    def cluster_applications(self, state):
-        return self.get_url("/cluster/apps", state=state)
+    def cluster_applications(self, *state):
+        """Gets information about YARN apps with the given state.
+
+        Parameters
+        ----------
+        *state: list, optional
```

Comment:
Two things on this one

1. All of the elements of `*args` need to be strings. That's probably worth documenting
2. If you pass a list in this will blow up:

```python
def foo(*args):
    print(','.join(args))

foo(['a'])
```

produces

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-c1210e32b7eb> in <module>()
      2     print(','.join(args))
      3 
----> 4 foo(['a'])

<ipython-input-23-c1210e32b7eb> in foo(*args)
      1 def foo(*args):
----> 2     print(','.join(args))
      3 
      4 foo(['a'])

TypeError: sequence item 0: expected str instance, list found
```

Reply:
Oops. The required type is str, not list. I'll fix the docstring.